request_library.md

```diff
@@ -0,0 +1,223 @@
+pip install -U requests
+pip install -U robotframework-requests
+```
+
+Now we can prepare RPI image. Install `RASPBIAN STRETCH WITH DESKTOP`. It is
+very good solution, because we can controll also GPIO's using buttons.
+
+![](web_browser_control.png)
+
+I hope that instruction of installation process is exact, so I will not duplicate
+the description. Take a few minutes for preparation of configurations for pins
+and restart gpio server:
+
+```
+sudo service gpio-server restart
+```
+
+For running server go to Pi-GPIO-Server and type:
+
+```
+python pi_gpio_server.py
+```
+
+Now, we can start writting code for control GPIO using Rest API and Robot framework.
+
+```
+*** Settings ***
+
+*** Variables ***
+
+*** Test Cases ***
+
+*** Keywords ***
+```
+
+Run RPi and check its IP address. We will use IP for connection with API server.
+My RPI has IP: 192.168.0.46. Documentation says that server runs on `5000` port. 
+So we can prepare URL variable. On `Variables` section type:
+
+```
+${URL} http://192.168.0.46:5000
+```
+
+Let's test out first test `Get signals`. Firstly  we should create session 
+for request control in Robot Framework. This convention uses all
+libraries eg. SSH.
+
+For create session we use keyword `Create session` with parameters `alias`, `URl`
+and optional `verify`.
+
+```
+*** Test Cases ***
+
+Get signal
+Create Session gpio_server ${URL} verify=True
+```
+
+We will use `alias` for all keywords which work on URL. All sessions should be
+closed using keyword `Delete All Sessions`.
+Next we need header, which should be send with request. One of the methods for creating
+header looks like:
+
+```
+Create Dictionary Content-Type=application/json Accept=application/json
+```
+
+Good trick is assigning headers to variable. For checking value of GPIO is prepared
+GET method. Documentation tells:
+
+```
+Read a single pin
+
+GET: /api/v1/pin/:num
+```
+
+Let's implement it in another way:
+
+```
+${pin}= Get Request gpio_server /api/v1/pin/18 ${headers}
+```
+
+As I said, `alias` is very important. Unfortunatelly for observing results
+Robot Framework has special keyword. Its name is `Log`.
+
+Whole test look like this:
+
+```
+*** Settings ***
+Library RequestsLibrary
+Library Collections
+
+
+*** Variables ***
+${URL} http://192.168.0.46:5000
+
+
+*** Test Cases ***
+
+Get signal
+Create Session gpio_server ${URL} verify=True
+${headers}= Create Dictionary Content-Type=application/json Accept=application/json
+${pin}= Get Request gpio_server /api/v1/pin/18 ${headers}
+Log  ${pin.json()}
+Delete All Sessions
+```
+
+For getting all signals you can only delete number of pin on hyperlink.
+Example:
+
+```
+Get signals
+Create Session gpio_server ${URL} verify=True
+${headers}= Create Dictionary Content-Type=application/json Accept=application/json
+${pinout}= Get Request gpio_server /api/v1/pin ${headers}
+Log ${pinout.json()}
+```
+
+That was easier part of our script.
+
+To set values for GPIO's is designed patch method.
```

Comment:
To set the values for the GPIOs, the patch method needs to be designed (?) - nie do końca rozumiem czy o to chodziło